From 70f9d5497bea15a1567c8c23d48cd48db9e4fe40 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Fri, 5 Jul 2024 09:08:28 +0200
Subject: [PATCH 1185/1695] plat-stm32mp1: move console_init() to init section

Move console_init() platform function from unpaged area to the
init area to optimize pager resisdent memory footprint.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 290a64ad34621cc713692a480112e0803f6b27fb plat-stm32mp1: move console_init() to init section]
Change-Id: Ica0ee2413b15bbba37f809252532183f8bb6003b
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/392003
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 core/arch/arm/plat-stm32mp1/link_dummies_init.c | 10 ++++++++++
 core/arch/arm/plat-stm32mp1/main.c              |  6 +++++-
 core/arch/arm/plat-stm32mp1/sub.mk              |  2 +-
 3 files changed, 16 insertions(+), 2 deletions(-)
 create mode 100644 core/arch/arm/plat-stm32mp1/link_dummies_init.c

diff --git a/core/arch/arm/plat-stm32mp1/link_dummies_init.c b/core/arch/arm/plat-stm32mp1/link_dummies_init.c
new file mode 100644
index 000000000..746f75840
--- /dev/null
+++ b/core/arch/arm/plat-stm32mp1/link_dummies_init.c
@@ -0,0 +1,10 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022, Linaro Limited
+ */
+#include <compiler.h>
+#include <console.h>
+
+void __section(".text.dummy.console_init") console_init(void)
+{
+}
diff --git a/core/arch/arm/plat-stm32mp1/main.c b/core/arch/arm/plat-stm32mp1/main.c
index 0338ebc0a..c43f1099b 100644
--- a/core/arch/arm/plat-stm32mp1/main.c
+++ b/core/arch/arm/plat-stm32mp1/main.c
@@ -96,7 +96,11 @@ service_init(platform_banner);
  */
 static struct stm32_uart_pdata console_data;
 
-void console_init(void)
+/*
+ * Note: this function is weak just to make it possible to exclude it from
+ * the unpaged area.
+ */
+void __weak console_init(void)
 {
 	/* Early console initialization before MMU setup */
 	struct uart {
diff --git a/core/arch/arm/plat-stm32mp1/sub.mk b/core/arch/arm/plat-stm32mp1/sub.mk
index 7c7894212..2db32800c 100644
--- a/core/arch/arm/plat-stm32mp1/sub.mk
+++ b/core/arch/arm/plat-stm32mp1/sub.mk
@@ -7,7 +7,7 @@ srcs-$(CFG_SCMI_MSG_DRIVERS) += scmi_server.c
 srcs-$(CFG_SCMI_SCPFW) += scmi_server_scpfw.c
 srcs-$(CFG_STM32MP1_SHARED_RESOURCES) += shared_resources.c
 srcs-$(CFG_TZC400) += plat_tzc400.c
-srcs-$(CFG_WITH_PAGER) += link_dummies_paged.c
+srcs-$(CFG_WITH_PAGER) += link_dummies_paged.c link_dummies_init.c
 
 subdirs-y += drivers
 subdirs-y += nsec-service
-- 
2.47.2

